### ISSUE TITLE
@_silgen_name attribute doesn't work on initializers

### ISSUE THREAD

**Issue by glessard** [apple/swift#68433]
Steps:
1. $ mkdir /tmp/silgenname && cd /tmp/silgenname && swift package init --type=library
2. append `type: .dynamic,` to line 11 of "Package.swift"
3. replace the contents of "Sources/silgenname/silgenname.swift" with
```
public struct Struct {
  var property: Int

  @_silgen_name("yeayeayea")
  mutating public func blah(_ v: Int) {
    property = v
  }

  @_silgen_name("whawhawha")
  public init(_ v: Int) {
    property = v
  }
}
```
4. $ swift build
5. $ nm -g `find . -name libsilgenname.dylib` | grep yeayeayea
There's a hit as expected
6. $ nm -g `find . -name libsilgenname.dylib` | grep whawhawha
Nothing. However, there is a symbol _$s10silgenname6StructVyACSicfC
7. $ swift demangle _\$s10silgenname6StructVyACSicfC
_$s10silgenname6StructVyACSicfC ---> silgenname.Struct.init(Swift.Int) -> silgenname.Struct
-> but that's the initializer we tried to _silgen_name.

It might be intended that @_silgen_name doesn't work on initializers, but it's not documented as such in the UnderscoredAttributes.md reference or the StandardLibraryProgrammersManual.md.

**Comment by jckarter** [comment]
Class initializers have multiple symbol names associated with them, so it isn't clear in the general case which symbol(s)' names would be affected by the attribute.

**Comment by glessard** [comment]
What about the struct case, though?

**Comment by glessard** [comment]
rdar://115296219

**Comment by glessard** [comment]
@DougGregor fixed this in d19f082665489d0a49ef3d201907bf7f68b3dae2 (https://github.com/apple/swift/pull/70907)
